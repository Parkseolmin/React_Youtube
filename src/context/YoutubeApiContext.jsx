import { FakeYoutube } from 'api/fakeYoutube';
import { Youtube } from 'api/youtube';
import { createContext, useContext } from 'react';

export const YoutubeApiContext = createContext();

// const youtube = new FakeYoutube();
const youtube = new Youtube();
export function YoutubeApiProvider({ children }) {
  return (
    <YoutubeApiContext.Provider value={{ youtube }}>
      {children}
    </YoutubeApiContext.Provider>
  );
}

export function useYoutubeApi() {
  return useContext(YoutubeApiContext);
}
