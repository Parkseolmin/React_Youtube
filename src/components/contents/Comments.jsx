import Loading from './Loading';
import NotFound from './NotFound';
import { formatAgo } from 'util/date';
import { useYoutubeApi } from 'context/YoutubeApiContext';
import { useInfiniteQuery } from '@tanstack/react-query';

export default function Comments({ videoId }) {
  const { youtube } = useYoutubeApi();
  const {
    data: comments,
    fetchNextPage,
    hasNextPage,
    isFetchingNextPage,
    isLoading,
    error,
  } = useInfiniteQuery({
    queryKey: ['comments', videoId],
    queryFn: async ({ pageParam = '' }) => {
      return youtube.commentsAPI(videoId, pageParam);
    },
    getNextPageParam: (lastPage) => lastPage.nextPageToken || undefined,
  });
  // const {
  //   data: comments,
  //   fetchNextPage,
  //   hasNextPage,
  //   isFetchingNextPage,
  //   isLoading,
  //   error,
  // } = useYoutubeInfiniteQuery(['comments', videoId], (youtube, pageParam) =>
  //   youtube.commentsAPI(videoId, pageParam)
  // );
  if (isLoading) {
    return <Loading />;
  }

  if (error || !comments) {
    return <NotFound />;
  }

  // HTML 태그 제거 함수
  const stripHtmlTags = (str) => {
    return str.replace(/<[^>]*>/g, '');
  };

  console.log('comments', comments);
  return (
    <div className='sm:p-3 lg:p-5 '>
      <h2>댓글</h2>
      {comments && (
        <ul>
          {comments.pages.map((page, pageIndex) =>
            page.items.map((item) => (
              <li
                key={`${pageIndex}-${item.id}`}
                className='flex gap-2 line-clamp-2 mt-3'
              >
                <img
                  className='w-6 h-6 rounded-full'
                  src={
                    item.snippet.topLevelComment.snippet.authorProfileImageUrl
                  }
                  alt='profileimage'
                />
                <div className='flex flex-col'>
                  <span className='flex gap-3 items-center'>
                    <p className='text-sm opacity-85 mb-2'>
                      {item.snippet.topLevelComment.snippet.authorDisplayName}
                    </p>
                    <p className='text-xs opacity-75'>
                      {formatAgo(
                        item.snippet.topLevelComment.snippet.publishedAt
                      )}
                    </p>
                  </span>
                  <span>
                    {stripHtmlTags(
                      item.snippet.topLevelComment.snippet.textDisplay
                    )}
                  </span>
                </div>
              </li>
            ))
          )}
        </ul>
      )}
      <div className='video__more'>
        {hasNextPage && (
          <button onClick={() => fetchNextPage()} disabled={isFetchingNextPage}>
            {isFetchingNextPage ? 'Loading more...' : '더보기'}
          </button>
        )}
      </div>
    </div>
  );
}
