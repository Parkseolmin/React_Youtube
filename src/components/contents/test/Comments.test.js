import { render, screen, waitFor } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { useYoutubeApi } from 'context/YoutubeApiContext';
import { MemoryRouter } from 'react-router-dom';
import Comments from '../Comments';

// youtube API 모킹
jest.mock('context/YoutubeApiContext');

// QueryClient 생성 함수
const createTestQueryClient = () =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false, // 재시도 방지
      },
    },
  });

describe('Comments Component', () => {
  const mockYoutube = {
    commentsAPI: jest.fn(),
  };

  beforeEach(() => {
    useYoutubeApi.mockReturnValue({ youtube: mockYoutube });
  });

  it('renders loading state', () => {
    mockYoutube.commentsAPI.mockReturnValue(new Promise(() => {})); // 대기 중 상태

    const { asFragment } = render(
      <QueryClientProvider client={createTestQueryClient()}>
        <MemoryRouter>
          <Comments videoId='123' />
        </MemoryRouter>
      </QueryClientProvider>
    );

    // Loading 컴포넌트가 렌더링되는지 확인
    expect(screen.getByText('로딩 중...')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders NotFound when there is an error', async () => {
    mockYoutube.commentsAPI.mockRejectedValue(
      new Error('Error fetching comments')
    );

    render(
      <QueryClientProvider client={createTestQueryClient()}>
        <MemoryRouter>
          <Comments videoId='123' />
        </MemoryRouter>
      </QueryClientProvider>
    );

    // NotFound 컴포넌트가 렌더링되는지 확인
    await waitFor(() => {
      expect(screen.getByText(/여기는 어떻게 오셨나요?/i)).toBeInTheDocument();
    });
  });

  it('renders comments correctly', async () => {
    const mockComments = {
      nextPageToken: 'next-token',
      items: [
        {
          id: '1',
          snippet: {
            topLevelComment: {
              snippet: {
                authorProfileImageUrl: 'https://example.com/image.jpg',
                authorDisplayName: 'John Doe',
                publishedAt: '2024-01-01T00:00:00Z',
                textDisplay: '<p>Hello, world!</p>',
              },
            },
          },
        },
      ],
    };

    mockYoutube.commentsAPI.mockResolvedValue({
      nextPageToken: 'next-token',
      items: mockComments.items,
    });

    const { asFragment } = render(
      <QueryClientProvider client={createTestQueryClient()}>
        <MemoryRouter>
          <Comments videoId='123' />
        </MemoryRouter>
      </QueryClientProvider>
    );

    // 댓글이 렌더링되는지 확인
    await waitFor(() => {
      expect(screen.getByText('John Doe')).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders more comments button when there are more comments', async () => {
    const mockComments = {
      nextPageToken: 'next-token',
      items: [
        {
          id: '1',
          snippet: {
            topLevelComment: {
              snippet: {
                authorProfileImageUrl: 'https://example.com/image.jpg',
                authorDisplayName: 'John Doe',
                publishedAt: '2024-01-01T00:00:00Z',
                textDisplay: '<p>Hello, world!</p>',
              },
            },
          },
        },
      ],
    };

    mockYoutube.commentsAPI.mockResolvedValue(mockComments);

    render(
      <QueryClientProvider client={createTestQueryClient()}>
        <MemoryRouter>
          <Comments videoId='123' />
        </MemoryRouter>
      </QueryClientProvider>
    );

    // 더보기 버튼이 렌더링되는지 확인
    await waitFor(() => {
      expect(screen.getByText('더보기')).toBeInTheDocument();
    });
  });
});
