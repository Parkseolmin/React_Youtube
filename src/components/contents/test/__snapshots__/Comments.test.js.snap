// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments Component renders comments correctly 1`] = `
<DocumentFragment>
  <div
    class="sm:p-3 lg:p-5 "
  >
    <h2>
      댓글
    </h2>
    <ul>
      <li
        class="flex gap-2 line-clamp-2 mt-3"
      >
        <img
          alt="profileimage"
          class="w-6 h-6 rounded-full"
          src="https://example.com/image.jpg"
        />
        <div
          class="flex flex-col"
        >
          <span
            class="flex gap-3 items-center"
          >
            <p
              class="text-sm opacity-85 mb-2"
            >
              John Doe
            </p>
            <p
              class="text-xs opacity-75"
            >
              8 months ago
            </p>
          </span>
          <span>
            Hello, world!
          </span>
        </div>
      </li>
    </ul>
    <div
      class="video__more"
    >
      <button>
        더보기
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Comments Component renders loading state 1`] = `
<DocumentFragment>
  <div
    class="loading-container"
  >
    <div
      class="loading-spinner"
    />
    <p
      class="loading-text"
    >
      로딩 중...
    </p>
  </div>
</DocumentFragment>
`;
